File: 103_Binary_Tree_ZigZag_Level_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>>vec;
    void traverse(TreeNode *root, int level) {
        if (root == NULL) return;
        if(level==vec.size())
            vec.push_back(vector<int> ());
        if(!(level&1)) {
            vec[level].push_back(root->val);
        } else {
            vec[level].insert(vec[level].begin(), root->val);
        }
        traverse(root->left, level+1);
        traverse(root->right, level+1);
    }
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        traverse(root,0);
        return vec;
    }
};

File: 103_Binary_Tree_ZigZag_Level_order_traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        q = [root]
        toggle = False
        output = []
        while q:
            out_arr = [x.val for x in q]
            if toggle: out_arr.reverse()
            output.append(out_arr)
            prev_q, q = q, []
            for i in range(0, len(prev_q)):
                elem = prev_q[i]
                if elem.left: q.append(elem.left)
                if elem.right: q.append(elem.right)
            toggle = not toggle
        return output
        